OSDOCS-14601 [NETOBSERV-2053] Flowlogs Flp filtering (#93270)
openshift/openshift-docs: modules/network-observability-config-flows-at-ingestion.adoc

[id="network-observability-config-flows-at-ingestion_{context}"]

= Filtering network flows at ingestion

You can create filters to limit the number of generated network flows, which is a way to reduce the amount of resources used by Network Observability.

There are two different kinds of filters that you can configure:

[id="ebpf-agent-filters_{context}"]
== eBPF agent filters

eBPF agent filters maximize performance because they take effect at the earliest stage of the Network Observability process

To configure eBPF agent filters with the Network Observability Operator, see "Filtering eBPF flow data using multiple rules".
//Filtering eBPF flow data using multiple rules is handled by PR 92375. Add as Additional resource in assembly

[id="flowlogs-pipeline-filters_{context}"]
== Flowlogs-pipeline filters

Flowlogs-pipeline filters use a simple query language to filter network flows:

[source, terminal]
----
(srcnamespace="netobserv" OR (srcnamespace="ingress" AND dstnamespace="netobserv")) AND srckind!="service"
----

Flowlogs-pipeline filters offer more filter flexibility because they take effect later in the process, but they primarily improve data storage.

> You can configure the Network Observability Operator in the `FlowCollector` resource's `spec.processor.filters.` parameter. For example:

Reviewer: ```suggestion
You can configure the `FlowCollector` resource's `spec.processor.filters.` parameter. For example:
```

.Example YAML Flowlogs-pipeline filter
[source,yaml]
----
apiVersion: flows.netobserv.io/v1beta2
kind: FlowCollector
metadata:
  name: cluster
spec:
  namespace: netobserv

>   deploymentModel: Direct

Reviewer: ```suggestion
```

  agent:
  processor:
    filters:
      - query: |
          (SrcK8S_Namespace="netobserv" OR (SrcK8S_Namespace="openshift-ingress" AND DstK8S_Namespace="netobserv"))
        outputTarget: Loki  <1>
        sampling: 10  <2>
----

> <1> The filter can target a specific output (such as Loki logs or exported data), or all outputs.

Reviewer: ```suggestion
<1> The filter can target a specific output (such as Loki logs, prometheus or exported data), or all outputs.
```

////
Syntax explanation:
Logical boolean operators (case insensitive)
and
or
Comparison operators
equals =
not equals !=
matches regexp =~
not matches regexp !~
greater than (or equal) > / >=
less than (or equal) < / <=
Unary operations
field is present: with(field)
field is absent: without(field)
Parenthesis-based priority
////

You can configure flowlogs-pipeline filters in the `spec.processor.filters` parameter of the `FlowCollector` resource. For example:

.Example YAML Flowlogs-pipeline filter
[source,yaml]
----
apiVersion: flows.netobserv.io/v1beta2
kind: FlowCollector
metadata:
  name: cluster
spec:
  namespace: netobserv
  agent:
  processor:
    filters:
      - query: |
          (SrcK8S_Namespace="netobserv" OR (SrcK8S_Namespace="openshift-ingress" AND DstK8S_Namespace="netobserv"))
        outputTarget: Loki  <1>
        sampling: 10  <2>
----
<1> Sends matching flows to a specific output, such as Loki, Prometheus, or an external system.

> <2> Optional. Applies a sampling ratio to limit the number of flows that match the query.

Reviewer: Need more explanation about sampling. Will let @jotak put it into words

Reviewer 2: maybe adding a short example like:

```suggestion
<2> Optional. Applies a sampling ratio to limit the number of matching flows to be stored or exported. For example, a value of 10 means 1 matching flow every 10 is kept.
```

// Module included in the following assemblies:

// * networking/network_observability/configuring-operators.adoc

:_mod-docs-content-type: CONCEPT
[id="network-observability-config-flows-at-ingestion_{context}"]

= Filtering network flows at ingestion

> You can create filters to reduce the number of generated network flows. Filtering network flows can help reduce the resource usage of the Network Observability Operator.

Reviewer 2: I wouldn't say "operator" here, it's more the other underlying components that are affected, not the operator itself.

```suggestion
You can create filters to reduce the number of generated network flows. Filtering network flows can help reduce the resource usage of the Network Observability components.
```

There are two different kinds of filters that you can configure:

[id="ebpf-agent-filters_{context}"]
== eBPF agent filters

> eBPF agent filters maximize performance because they take effect at the earliest stage of the Network Observability Operator process.

Reviewer 2: same here, it's not about the operator.
I'd suggest this:
```suggestion
eBPF agent filters maximize performance because they take effect at the earliest stage of the network flows collection process.
```

> To configure eBPF agent filters with the Network Observability Operator, see "Filtering eBPF flow data using multiple rules".

Author: Should most of the content from "Observing" be moved to "Configuring" then? 

Working with conversion tracking
Working with packet drops
Working with DNS tracking
Working with RTT tracing

And then by extension, all the Conceptual content for those would likely also need to move. 

Conceptual content like this: this https://93270--ocpdocs-pr.netlify.app/openshift-enterprise/latest/observability/network_observability/observing-network-traffic#network-observability-ebpf-flow-rule-filter_nw-observe-network-traffic 

Seems like maybe a re-thinking and discussion of what is considered "Observability" content and what is considered "Configuring" content is needed. That is outside the scope of this PR, so for now, I'm going to leave it as it is, and created this Jira https://issues.redhat.com/browse/OSDOCS-14779 to review the "Observing the network traffic from the Overview view" content and the "Configuring the Network Observability Operator" content, and see what/how to move things around.

> To configure eBPF agent filters with the Network Observability Operator, see "Filtering eBPF flow data using multiple rules".

Reviewer 2: I didn't mean to start such a big refactoring, it's really the filtering section that is pure config. The other content that you mention _do_ tell about what changes in the UI. Not the ebpf filtering content.

[id="flowlogs-pipeline-filters_{context}"]
== Flowlogs-pipeline filters

Flowlogs-pipeline filters provide greater control over traffic selection because they take effect later in the Network Observability process. They are primarily used to improve data storage.

[id="flowlogs-pipeline-filters-query-language_{context}"]
=== Flowlogs-pipeline filters query language

Flowlogs-pipeline filters use a simple query language to filter network flows:

[source]
----
(srcnamespace="netobserv" OR (srcnamespace="ingress" AND dstnamespace="netobserv")) AND srckind!="service"
----

The query language uses the following syntax:

Logical boolean operators (not case-sensitive)::
* `and`
* `or`
Comparison operators::
* equals ``=`
* not equals `!=`
* matches regexp `=~`
* not matches regexp `!~`
* less than (or equal) `<` / `<=`
* greater than (or equal) `>` / `>=`
Unary operations::
* field is present: `with(field)`
* field is absent: `without(field)`
Parenthesis-based priority::
{empty}

[id="flowlogs-pipeline-configure_{context}"]
=== Flowlogs-pipeline configuration

> You can configure flowlogs-pipeline filters in the `spec.processor.filters` parameter of the `FlowCollector` resource. For example:

Reviewer: ```suggestion
You can configure flowlogs-pipeline filters in the `spec.processor.filters`section of the `FlowCollector` resource. For example:
```
